pkg/iam/validator/jwt.go

```diff
@@ -0,0 +1,228 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package validator
+
+import (
+	"crypto"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	jwtgo "github.com/dgrijalva/jwt-go"
+	xnet "github.com/minio/minio/pkg/net"
+)
+
+// JWKSArgs - RSA authentication target arguments
+type JWKSArgs struct {
+	URL       *xnet.URL `json:"url"`
+	publicKey crypto.PublicKey
+}
+
+// Validate JWT authentication target arguments
+func (r *JWKSArgs) Validate() error {
+	return nil
+}
+
+// PopulatePublicKey - populates a new publickey from the JWKS URL.
+func (r *JWKSArgs) PopulatePublicKey() error {
+	insecureClient := &http.Client{Transport: newCustomHTTPTransport(true)}
+	client := &http.Client{Transport: newCustomHTTPTransport(false)}
+	resp, err := client.Get(r.URL.String())
+	if err != nil {
+		resp, err = insecureClient.Get(r.URL.String())
+		if err != nil {
+			return err
+		}
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return errors.New(resp.Status)
+	}
+
+	var jwk JWKS
+	if err = json.NewDecoder(resp.Body).Decode(&jwk); err != nil {
+		return err
+	}
+
+	r.publicKey, err = jwk.Keys[0].DecodePublicKey()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UnmarshalJSON - decodes JSON data.
+func (r *JWKSArgs) UnmarshalJSON(data []byte) error {
+	// subtype to avoid recursive call to UnmarshalJSON()
+	type subJWKSArgs JWKSArgs
+	var sr subJWKSArgs
+
+	// IAM related envs.
+	if jwksURL, ok := os.LookupEnv("MINIO_IAM_JWKS_URL"); ok {
+		u, err := xnet.ParseURL(jwksURL)
+		if err != nil {
+			return err
+		}
+		sr.URL = u
+	} else {
+		if err := json.Unmarshal(data, &sr); err != nil {
+			return err
+		}
+	}
+
+	ar := JWKSArgs(sr)
+	if ar.URL == nil || ar.URL.String() == "" {
+		*r = ar
+		return nil
+	}
+	if err := ar.Validate(); err != nil {
+		return err
+	}
+
+	if err := ar.PopulatePublicKey(); err != nil {
+		return err
+	}
+
+	*r = ar
+	return nil
+}
+
+// JWT - rs client grants provider details.
+type JWT struct {
+	args JWKSArgs
+}
+
+func expToInt64(expI interface{}) (expAt int64, err error) {
+	switch exp := expI.(type) {
+	case float64:
+		expAt = int64(exp)
+	case int64:
+		expAt = exp
+	case json.Number:
+		expAt, err = exp.Int64()
+		if err != nil {
+			return 0, err
+		}
+	default:
+		return 0, errors.New("invalid expiry value")
+	}
+	return expAt, nil
+}
+
+func getDefaultExpiration(dsecs string) (time.Duration, error) {
+	defaultExpiryDuration := time.Duration(60) * time.Minute // Defaults to 1hr.
+	if dsecs != "" {
+		expirySecs, err := strconv.ParseInt(dsecs, 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		// The duration, in seconds, of the role session.
+		// The value can range from 900 seconds (15 minutes)
+		// to 12 hours.
+		if expirySecs < 900 || expirySecs > 43200 {
```

Comment:
If I put in 60 seconds when settings OAuth config in WSO2, this will return an error, is this really needed ?

Reply:
It will only return error if DurationSeconds parameter is < 900 it has nothing to do with WSO2 config.

Reply:
![screenshot from 2018-10-09 17-23-53](https://user-images.githubusercontent.com/283197/46683520-509b3400-cbe8-11e8-9162-b5734174dfaf.png)

This will show the following error
```
API: AssumeRoleWithClientGrants()
Time: 17:23:20 CET 10/09/2018
RequestID: 155BFD4C02BD5D79
RemoteHost: 127.0.0.1
UserAgent: Go-http-client/1.1
Error: out of range value for duration in seconds
       1: cmd/logger/logger.go:294:logger.LogIf()
       2: cmd/sts-handlers.go:170:cmd.(*stsAPIHandlers).AssumeRoleWithClientGrants()
       3: cmd/sts-handlers.go:49:cmd.(*stsAPIHandlers).AssumeRoleWithClientGrants-fm()
       4: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
```

Reply:
> This is JWT validation @vadmeste and it returns an error correctly, this API can never support creds which expire in 60 secs. Its compatible with the AWS implementation of WebIdentity.

Yes, right I understand now